local_planner/src/nodes/waypoint_generator.cpp

```diff
@@ -260,101 +260,49 @@ void WaypointGenerator::nextSmoothYaw(float dt) {
-    if (planner_info_.reach_altitude && !reached_goal_) {
-      int ind_dist = 100;
-      int i = 0;
-      for (std::vector<int>::iterator it = z_FOV_idx_.begin();
-           it != z_FOV_idx_.end(); ++it) {
-        if (std::abs(z_FOV_idx_[i] - p_index.x()) < ind_dist) {
-          ind_dist = std::abs(z_FOV_idx_[i] - p_index.x());
-        }
-        i++;
-      }
-      float angle_diff = static_cast<float>(std::abs(ALPHA_RES * ind_dist));
-      float hover_angle = 30.0f;
-      angle_diff = std::min(angle_diff, hover_angle);
-      speed_ = speed_ * (1.0f - angle_diff / hover_angle);
-    }
-  }
-  velocity_time_ = getSystemTime();
-
-  // calculate correction for computation delay
-  float since_update_sec =
-      static_cast<float>((getSystemTime() - update_time_).toSec());
-  speed_ += (since_update_sec * curr_vel_magnitude_);
-
-  // break before goal: if the vehicle is closer to the goal than a velocity
-  // dependent distance, the speed is limited
-  Eigen::Vector3f pos_to_goal =
-      (goal_ - toEigen(pose_.pose.position)).cwiseAbs();
-
-  if (pos_to_goal.x() < param_.factor_close_to_goal_start_speed_limitation *
-                            curr_vel_magnitude_ &&
-      pos_to_goal.y() < param_.factor_close_to_goal_start_speed_limitation *
-                            curr_vel_magnitude_) {
-    limit_speed_close_to_goal_ = true;
-  } else if (pos_to_goal.x() >
-                 param_.factor_close_to_goal_stop_speed_limitation *
-                     curr_vel_magnitude_ ||
-             pos_to_goal.y() >
-                 param_.factor_close_to_goal_stop_speed_limitation *
-                     curr_vel_magnitude_) {
-    limit_speed_close_to_goal_ = false;
-  }
-  if (limit_speed_close_to_goal_) {
-    speed_ = std::min(
-        speed_, param_.max_speed_close_to_goal_factor * pos_to_goal.norm());
-    speed_ = std::max(speed_, param_.min_speed_close_to_goal);
-  }
-
-  // set waypoint to correct speed
+  float angle_diff_deg =
+      std::abs(nextYaw(pose_, output_.goto_position) - curr_yaw_) * 180.f /
+      M_PI_F;
+  angle_diff_deg = std::min(h_FOV_ / 2, angle_diff_deg);  // Clamp at h_FOV/2
+  speed_ =
+      speed_ * (1.0f - 2 * angle_diff_deg / h_FOV_);  // throttle if outside FOV
+
+  // Scale the pose_to_wp by the speed
   Eigen::Vector3f pose_to_wp =
-      toEigen(output_.adapted_goto_position) - toEigen(pose_.pose.position);
-  if (pose_to_wp.norm() > 0.01f) pose_to_wp.normalize();
-
+      toEigen(output_.goto_position) - toEigen(pose_.pose.position);
+  Eigen::Vector3f pose_to_goal = goal_ - toEigen(pose_.pose.position);
+  if (pose_to_wp.norm() > 0.1f) pose_to_wp.normalize();
   pose_to_wp *= speed_;
 
-  output_.adapted_goto_position =
-      toPoint(toEigen(pose_.pose.position) + pose_to_wp);
+  // If the waypoint is in the direction of the goal, but farther, clamp to goal
+  if (std::atan2(pose_to_wp.cross(pose_to_goal).norm(),
+                 pose_to_wp.dot(pose_to_goal)) < 0.01f &&
```

Comment:
I've thought about it, and I guess that's usually safe, as long as we keep path segments short.

It's not completely obvious though, because there could be a path around an obstacle that is in front of the drone (thus dot product of direct-line-to-goal with path is positive) and still has the goal closer than the speed-adapted waypoint.